File: python/open_guardrail/guards/config_key_detect.py
"""Detect configuration keys and values in output."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"connection_string\s*[=:]\s*[\"']?"
        r"[^\s\"']+",
        re.IGNORECASE,
    ),
    re.compile(
        r"database_url\s*[=:]\s*[\"']?"
        r"[^\s\"']+",
        re.IGNORECASE,
    ),
    re.compile(
        r"redis_url\s*[=:]\s*[\"']?[^\s\"']+",
        re.IGNORECASE,
    ),
    re.compile(
        r"smtp_(host|password|user)\s*[=:]\s*"
        r"[\"']?[^\s\"']+",
        re.IGNORECASE,
    ),
    re.compile(
        r"config\.(get|set)\s*\(\s*[\"']"
        r"[^\"']*secret[^\"']*[\"']",
        re.IGNORECASE,
    ),
    re.compile(
        r"DB_(HOST|PASSWORD|USER|NAME)\s*[=:]"
        r"\s*[\"']?[^\s\"']+",
        re.IGNORECASE,
    ),
    re.compile(
        r"MONGO_URI\s*[=:]\s*[\"']?[^\s\"']+",
        re.IGNORECASE,
    ),
    re.compile(
        r"AWS_(ACCESS|SECRET|REGION)\s*[=:]"
        r"\s*[\"']?[^\s\"']+",
        re.IGNORECASE,
    ),
    re.compile(
        r"PRIVATE_KEY\s*[=:]\s*[\"']?[^\s\"']+",
        re.IGNORECASE,
    ),
    re.compile(
        r"api_secret\s*[=:]\s*[\"']?[^\s\"']+",
        re.IGNORECASE,
    ),
]


class _ConfigKeyDetect:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "config-key-detect"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="config-key-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Configuration key detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def config_key_detect(
    *, action: str = "block"
) -> _ConfigKeyDetect:
    return _ConfigKeyDetect(action=action)

File: packages/guards/src/config-key-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ConfigKeyDetectOptions {
  action: 'block' | 'warn';
}

const PATTERNS: RegExp[] = [
  /connection_string\s*[=:]\s*["']?[^\s"']+/i,
  /database_url\s*[=:]\s*["']?[^\s"']+/i,
  /redis_url\s*[=:]\s*["']?[^\s"']+/i,
  /smtp_(host|password|user)\s*[=:]\s*["']?[^\s"']+/i,
  /config\.(get|set)\s*\(\s*["'][^"']*secret[^"']*["']/i,
  /DB_(HOST|PASSWORD|USER|NAME)\s*[=:]\s*["']?[^\s"']+/i,
  /MONGO_URI\s*[=:]\s*["']?[^\s"']+/i,
  /AWS_(ACCESS|SECRET|REGION)\s*[=:]\s*["']?[^\s"']+/i,
  /PRIVATE_KEY\s*[=:]\s*["']?[^\s"']+/i,
  /api_secret\s*[=:]\s*["']?[^\s"']+/i,
];

export function configKeyDetect(options: ConfigKeyDetectOptions): Guard {
  return {
    name: 'config-key-detect',
    version: '0.1.0',
    description: 'Detects configuration keys and values in output',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'config-key-detect',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
